epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala

```diff
@@ -0,0 +1,300 @@
+                  go(a, after)
+                else
+                  IO.asyncCheckAttempt[Int] { cb =>
+                    IO {
+                      readCallback = cb
+                      // check again before we suspend
+                      val now = readReadyCounter
+                      if (now != before) {
+                        readCallback = null
+                        Right(now)
+                      } else {
+                        // block here until `readCallback` is invoked
+                        Left(Some(IO(this.readCallback = null)))
+                      }
+                    }
+                  }.flatMap(go(a, _))
+              }
+            case Right(b) => IO.pure(b)
+          }
+
+        IO(readReadyCounter).flatMap(go(a, _))
+      }
+
+    def pollWriteRec[A, B](a: A)(f: A => IO[Either[A, B]]): IO[B] =
+      writeMutex.lock.surround {
+        def go(a: A, before: Int): IO[B] =
+          f(a).flatMap {
+            case Left(a) =>
+              IO(writeReadyCounter).flatMap { after =>
+                if (before != after)
+                  // there was a write-ready notification since we started, try again immediately
+                  go(a, after)
+                else
+                  IO.asyncCheckAttempt[Int] { cb =>
+                    IO {
+                      writeCallback = cb
+                      // check again before we suspend
+                      val now = writeReadyCounter
+                      if (now != before) {
+                        writeCallback = null
+                        Right(now)
+                      } else Left(Some(IO(this.writeCallback = null)))
+                    }
+                  }.flatMap(go(a, _))
+              }
+            case Right(b) => IO.pure(b)
+          }
+
+        IO(writeReadyCounter).flatMap(go(a, _))
+      }
+  }
+
+  final class Poller private[EpollSystem] (epfd: Int, evfd: Int) {
+
+    // fixed array used only when reading from eventfd
+    private final lazy val readBuf = Memory.allocate(globalRuntime, 8)
+
+    // fixed array used only when writing to eventfd
+    private final lazy val writeBuf = byteToPtr(Array.apply[Byte](0, 0, 0, 0, 0, 0, 0, 1))
+
+    private[this] val handles: ConcurrentHashMap[Int, PollHandle] = new ConcurrentHashMap()
+
+    private[EpollSystem] def close(): Unit =
+      if (libc.jnr.close(epfd) != 0)
+        throw new IOException(strerror(errno()))
+
+    private[EpollSystem] def poll(timeout: Long): Boolean = {
+
+      val events = Memory.allocate(globalRuntime, EpollEvent.CStructSize * MaxEvents)
+      var polled = false
+
+      @tailrec
+      def processEvents(timeout: Int): Unit = {
+
+        val triggeredEvents = epoll_wait(epfd, events, MaxEvents, timeout)
+
+        if (triggeredEvents >= 0) {
+
+          var i = 0
+          while (i < triggeredEvents) {
+            val epollEvent =
+              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)
```

Comment:
If we pre-allocate and cache these I wonder if it will give any performance improvement.